SCMOD-13870: Added rabbit Exception handling

xbreizh opened the pull request:
https://portal.digitalsafe.net/browse/SCMOD-13870

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](http://sou-jenkins2.swinfra.net:8080/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=JobService&REPO_NAME=job-service&BRANCH_NAME=SCMOD-13870) to go ahead and create the build...

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/JobService/job/JobService~job-service~SCMOD-13870~CI

xbreizh commented:
Hi @dermot-hardy so here, I decomposed in several steps:
1- [Added retry on createConnection](https://github.com/JobService/job-service/pull/252/commits/8b14fbe1688f026fd0180606534c09e2c3fd4699)
2- Returning 200 if subtask reported in the database but failing to send a message -> not required anymore as we want to get rid of the sendMessage()
3- Fill job_task_data on each job creation -> done (needs refactoring).
4- Add scheduler function to pickup job_task_data data automatically
The dilemna that I have at the moment here, is that the [getJobCanBeProgressed()](https://github.com/JobService/job-service/blob/develop/job-service-db/src/main/resources/procedures/getJobCanBeProgressed.sql) checks whether a job is in that table in order to know if it's available to run, when with that new way of working, any new job without dependency would be in that table and available for running. So, in order to differentiate and filter, I added  a new column ["Type"](https://github.com/JobService/job-service/blob/3191789e7b76480a3ee3317d4520ccbe183d8b57/job-service-db/src/main/resources/v4.0.0/changelog.xml#L107) with default value as Depending, so passing "Not Depending" on job creation.
In order to have the passing within job-service-container, I added the [job-service-scheduled-executor](https://github.com/JobService/job-service/pull/252/commits/4c254c79b4b3ee8cff29b865c786e61bc11e45cf) and also [changed the build order](https://github.com/JobService/job-service/pull/252/commits/3191789e7b76480a3ee3317d4520ccbe183d8b57) so that it would be available to run.
With that move, we manage to keep the tests passing as before (I removed the sendMessage section from them so we have the behaviour that we seek)
There's only one resistance at the mn that I am [working on](https://github.com/JobService/job-service/blob/3191789e7b76480a3ee3317d4520ccbe183d8b57/job-service-container/src/test/java/com/hpe/caf/services/job/api/JobServiceIT.java#L740). I am unsure if we need to keep it and fix or just get rid of it. I commented it out for now and try to see if I can fix it
5- Remove reportFailure() and associated -> quick cleanup
TODO functionality to poke scheduler on job creation

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/JobService/job/JobService~job-service~SCMOD-13870~CI

buildmachine-sou-jenkins2 commented:
The Documentation QA site for this branch has been built:
https://pages.github.houston.softwaregrp.net/sepg-docs-qa/JobService_CI_job-service_SCMOD-13870

xbreizh commented:
@dermot-hardy About the build, it tends to randomly throw a EOFException:
`[ERROR] Failed to execute goal com.googlecode.maven-download-plugin:download-maven-plugin:1.3.0:wget (download-target-file) on project job-service-docs: IO Error: EOFException -> [Help 1]
`
I discussed it with @tonymcveigh and we 

> Could we implement a retry in case in requests it too quickly or decides it's EOF too quickly?
Also, 

> Are there other build doing something similar? Do they use wget?

If we insist a couple of times, the build finally passes, but it might be worth fixing it, either by spending some time on investigating the root source or by using a quick fix as described above. What do you think?

xbreizh commented:
I removed the extra column that I previously added.
I also temporarily added a Sleep on 2 of the tests in order to let the scheduler wake up and do its work
The next step is to implement a trigger on that scheduler method (pollDatabaseForJobsToRun()) on job creation (job without dependencies)

dermot-hardy commented on docs/pom.xml:
```diff
@@ -197,6 +197,7 @@
                             <overwrite>true</overwrite>
                             <skipCache>true</skipCache>
                             <url>http://${docker.host.address}:${webServerPort}/docs-site.tar.gz</url>
+                            <retries>10</retries>
```
This is nothing to do with this Jira.  Raise a separate Jira/PR for it if it is required.

xbreizh replied:
I added it here in an attempt to get the build stable wrt that above comment (we keep hitting job-service-docs: IO Error: EOFException) but indeed, unrelated to this jira so removed.

dermot-hardy commented on pom.xml:
```diff
@@ -247,11 +247,11 @@
         <module>job-service-html</module>
         <module>job-service-scheduled-executor</module>
         <module>job-service-scheduled-executor-container</module>
+        <module>job-service-container</module>
```
Why is this being moved?

xbreizh replied:
[from](https://github.com/JobService/job-service/pull/252#issuecomment-850564347) I added the job-service-scheduled-executor to be available in job-service-container's tests. So, I had to reverse the building order so it gets available there

dermot-hardy commented on job-service-container/src/test/java/com/hpe/caf/services/job/api/JobServiceIT.java:
```diff
@@ -737,7 +737,7 @@ public void testGetJobsCountFromDifferentPartition() throws ApiException {
      * @throws Exception
      */
-    @Test
+   // @Test
```
Why is this being commented out?

If it is no longer relevant then why are we keeping it around?  If it can be removed than does that allow any other functions (`testMessagesPutOnQueue` perhaps) to be removed?

xbreizh replied:
I think that this test is no longer relevant, but I wasn't 100% sure so wanted to check it with you. If we get rid of it, then yes testMessagesPutOnQueue() can be removed too.

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServicesFactory.java:
```diff
@@ -42,7 +45,14 @@
+        try {
+            //  Create connection and channel for publishing messages.
+            connection = createConnection(configuration);
+        } catch (IOException | TimeoutException e) {
```
```suggestion
        } catch (final IOException | TimeoutException e) {
```

dermot-hardy commented on job-service/src/test/java/com/hpe/caf/services/job/api/JobsPutTest.java:
```diff
@@ -145,7 +142,6 @@ public void setup() throws Exception {
         JobTypes.initialise(() -> Collections.singletonList(basicJobType));
 
         //  Mock QueueServices calls.
-        doNothing().when(mockQueueServices).sendMessage(any(), any(), any(), any(), anyBoolean());
```
Can the `anyBoolean()` import now be removed?

xbreizh replied:
Removed unused import

dermot-hardy commented on worker-jobtracking-container/src/test/java/com/hpe/caf/worker/jobtracking/JobDatabase.java:
```diff
@@ -26,6 +28,7 @@
 import java.text.MessageFormat;
 import java.time.Instant;
 import java.util.Date;
+import java.util.List;
```
Why is this being added?

xbreizh replied:
Removed unused import

dermot-hardy commented on worker-jobtracking-container/src/test/java/com/hpe/caf/worker/jobtracking/JobDatabase.java:
```diff
@@ -59,13 +62,24 @@ public void createJobTask(
+            CallableStatement stmt = connection.prepareCall("{call create_job(?,?,?,?,?,?,?,?,?,?,?,?,?)}")) {
+
             stmt.setString(1, partitionId);
-            stmt.setString(2, jobId);
-            stmt.setString(3, name);
-            stmt.setString(4, description);
-            stmt.setString(5, data);
-            stmt.setInt(6, jobHash);
```
Why are these lines changing?

xbreizh replied:
restored

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsGetById.java:
```diff
@@ -65,8 +65,10 @@ public static Job getJob(final String partitionId, String jobId) throws Exceptio
+            if (null != job) {
+                LOG.info("Job progress {}/{} status {}  progression {}", partitionId, jobId, job.getStatus(),
+                        job.getPercentageComplete());
+            }            return job;
```
Merge `develop` into this branch to get rid of this change.

xbreizh replied:
Done

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsPut.java:
```diff
@@ -75,6 +63,7 @@
      */
     public static String createOrUpdateJob(final String partitionId, String jobId, NewJob job) throws Exception {
         try {
+            final boolean hasDependencies;
```
Surprising to see this from a non-C developer (because in C variables have to be declared at the start of functions).  In modern languages they should be declared as close to their initial usage as is sensible (i.e. just above the `jobCreated` declaration in this case).

xbreizh replied:
Moved down.

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsPut.java:
```diff
@@ -188,46 +177,28 @@ public static String createOrUpdateJob(final String partitionId, String jobId, N
             }
 
-            if (!databaseHelper.canJobBeProgressed(partitionId, jobId)) {
+            if (hasDependencies && !databaseHelper.canJobBeProgressed(partitionId, jobId)) {
```
What is the purpose of this?  It looks like you are returning `"create"` either way.

xbreizh replied:
True, strangely we were doing it before and I just tried to adapt. I guess there was a use for it some time ago, but you're right, not anymore and it can be removed. Also, that's the only place in the code where we use canJobBeProgressed() so I cleaned that one too

xbreizh commented:
There is some correction required in the acceptance-tests that I can't figure at the moment.
The manual tests are working as expected -> we can see in the scheduler logs that we trigger it as soon as the job is created , whereas there's a delay when the cron runs.
I added some details in the existing logs to determine the source of the trigger (job creation or cron)

xbreizh commented:
It seems that the process got slightly slower, I had to keep some delay in a [test](https://github.com/JobService/job-service/blob/49187cdbcb7bcccc3a8fdf839e7471dfbea948d8/job-service-acceptance-tests/src/test/java/com/hpe/caf/jobservice/acceptance/JobServiceEndToEndIT.java#L711) to have it passing.
For it, I used a new library, [awaitability](http://www.awaitility.org/) so I can set a max value for the waiting duration instead of a fixed one. I imagine that in some cases (maybe not that exact one as the duration is minimal) it can help speeding up the tests?

dermot-hardy commented on job-service-acceptance-tests/pom.xml:
```diff
@@ -455,6 +455,46 @@
                             </run>
                         </image>
+    
+                        <!-- Run the Job Service Scheduled Executor -->
```
Why is this being moved?

xbreizh replied:
The idea was to have it created before the jobservice container. Moved it back down

dermot-hardy commented on job-service-container/pom.xml:
```diff
@@ -556,6 +556,35 @@
+                        <image>
+                            <alias>job-service-scheduled-executor</alias>
+                            <name>${dockerJobServiceOrg}job-service-scheduled-executor${dockerProjectVersion}</name>
+                            <run>
```
You should probably have a `wait` section in here also.

xbreizh replied:
Added

dermot-hardy commented on job-service-scheduled-executor-container/pom.xml:
```diff
@@ -171,6 +171,11 @@
+                                <env>
+                                    <CAF_WORKER_INPUT_QUEUE>jobservicescheduler-in</CAF_WORKER_INPUT_QUEUE>
+                                </env>
+                            </run>
```
This doesn't make any sense.  Did you mean to put the `env` element inside `build`, for it to become the default?

xbreizh replied:
Moved into build

dermot-hardy commented on job-service-acceptance-tests/pom.xml:
```diff
@@ -629,12 +630,15 @@
                             <name>${dockerJobServiceOrg}job-service-caller${dockerProjectVersion}</name>
                         </image>
-
+    
```
Pointless whitespace change

xbreizh replied:
Done

dermot-hardy commented on job-service-acceptance-tests/pom.xml:
```diff
@@ -656,9 +661,15 @@
+                                </wait>
                             </run>
                         </image>
-
```
Pointless whitespace change

dermot-hardy commented on job-service-config/src/main/java/com/hpe/caf/services/configuration/AppConfig.java:
```diff
@@ -15,16 +15,16 @@
 import java.util.Locale;
 import java.util.regex.Pattern;
 
+import org.springframework.beans.factory.annotation.Autowired;
```
Put the imports back where they were - moving them is not for this Jira

xbreizh replied:
Sorry, it's just the IDE ordering alphabetically. Done

dermot-hardy commented on job-service-config/src/main/java/com/hpe/caf/services/configuration/AppConfig.java:
```diff
@@ -120,4 +120,12 @@ public String getResumeJobQueue()
     {
         return environment.getProperty("CAF_JOB_SERVICE_RESUME_JOB_QUEUE");
     }
+    
```
```suggestion

```

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java:
```diff
@@ -68,6 +68,7 @@ public void sendMessage(
         final String partitionId, String jobId, WorkerAction workerAction, AppConfig config, final boolean includeTrackingInfo
     ) throws IOException, InterruptedException, TimeoutException
     {
+        LOG.info("sending message");
```
Do we need this?  Surely it will fire quite a lot and it doesn't tell us much.  If you think it should stay then should it be downgraded to debug, and should it include at least some details about the message being sent?

xbreizh replied:
No, that can go I think, I added it for debugging. Removed

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java:
```diff
@@ -122,7 +123,6 @@ public void sendMessage(
                 trackingInfo,
                 null,
                 MDC.get(MDC_KEY));
-
```
Put this blank line back.

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java:
```diff
@@ -141,9 +141,10 @@ public void sendMessage(
+    public void publishMessage(final byte[] taskMessageBytes)
             throws IOException, InterruptedException, TimeoutException
     {
+        LOG.info("target queue: {}", targetQueue);
```
Do we need this? Surely it will fire quite a lot and it doesn't tell us much. If you think it should stay then should it be downgraded to debug?

xbreizh replied:
Removed

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsPut.java:
```diff
@@ -162,13 +164,6 @@ public static String createOrUpdateJob(final String partitionId, String jobId, N
-
-            //  Get app config settings.
-            LOG.debug("createOrUpdateJob: Reading database and RabbitMQ connection properties...");
-            AppConfig config = AppConfigProvider.getAppConfigProperties();
```
Making these static will have an effect on what happens if there is an error.  Does it need to be done?

xbreizh replied:
The idea behind it was to help reducing the number of parameters passed between functions. I don't get what the downside would be in case of error? More risk of error if we call the function several times instead of storing into a parameter and pass it along?

dermot-hardy replied:
The problem is the context in which it is initialized.  By doing this you moved initialization from when the `createOrUpdateJob()` call is made to a time before any of the class methods are called (possibly even to service startup).  When it was when the `createOrUpdateJob()` call is made the error could be bubbled back to the client whereas that cannot happen during class initialization.  If you just want to reduce the parameters between functions without changing the initialization context then one way would be to use a nested class.  For example [`RandomSamplingWorker.Task`](https://github.houston.softwaregrp.net/Verity/worker-sampling/blob/v1.1.1/worker-sampling/src/main/java/com/microfocus/apollo/workers/sampling/RandomSamplingWorker.java#L101) takes that approach.  Note that the parameters are grouped into a nested object.  They are then used by [`process()`](https://github.houston.softwaregrp.net/Verity/worker-sampling/blob/v1.1.1/worker-sampling/src/main/java/com/microfocus/apollo/workers/sampling/RandomSamplingWorker.java#L141), [`getSamplingDispatchJobParametersDigest()`](https://github.houston.softwaregrp.net/Verity/worker-sampling/blob/v1.1.1/worker-sampling/src/main/java/com/microfocus/apollo/workers/sampling/RandomSamplingWorker.java#L232), [`createSamplingDispatchJobParameters()`](https://github.houston.softwaregrp.net/Verity/worker-sampling/blob/v1.1.1/worker-sampling/src/main/java/com/microfocus/apollo/workers/sampling/RandomSamplingWorker.java#L248), etc., all without being repeatedly passed around.

xbreizh replied:
Oh ok, thanks for the explanation.

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsPut.java:
```diff
@@ -190,44 +185,22 @@ public static String createOrUpdateJob(final String partitionId, String jobId, N
-
+    
+            triggerScheduler();
+    
```
```suggestion

            triggerScheduler();

```

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsPut.java:
```diff
@@ -236,17 +209,33 @@ public static String createOrUpdateJob(final String partitionId, String jobId, N
+                    UUID.randomUUID().toString(),
+                    DocumentWorkerConstants.DOCUMENT_TASK_NAME,
+                    1,
+                    OBJECT_MAPPER.writeValueAsBytes(new DocumentWorkerDocumentTask()),
```
Can we just use `codec.serialise()` for this?  I'm not sure why we're introducing `OBJECT_MAPPER`.

xbreizh replied:
replaced

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsPut.java:
```diff
@@ -262,16 +251,41 @@ private static void closeQueueConnection(final QueueServices queueServices)
     }
-
+    
+    private static byte[] serializingData(final Object taskDataObj)
```
You don't normally use the `-ing` form of the verb when naming methods.  Can this be changed to `serializeData`?

xbreizh replied:
Changed

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsPut.java:
```diff
@@ -236,17 +209,33 @@ public static String createOrUpdateJob(final String partitionId, String jobId, N
+            final byte[] taskMessageBytes = serializingData(taskMessage);
+            queueServices.publishMessage(taskMessageBytes);
+        } catch (final Exception ex) {
+           LOG.error("fail to ping the scheduler {}", ex.getMessage());
```
Downgrade to `WARN`.  Also start log statements with capital letters like a sentence, and use past tense where appropriate (i.e. `fail` --> `Failed`).

xbreizh replied:
Done

dermot-hardy commented on job-service/src/main/java/com/hpe/caf/services/job/api/JobsPut.java:
```diff
@@ -236,17 +209,33 @@ public static String createOrUpdateJob(final String partitionId, String jobId, N
-
-    private static byte[] getTaskDataBytes(final WorkerAction workerAction, final Codec codec)
+    
+    private static byte[] getTaskDataBytes(final WorkerAction workerAction)
```
Were the changes to this method necessary?

xbreizh replied:
Aiming to reduce the number of argument by making the codec global. Restoring

dermot-hardy commented on job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/ScheduledExecutor.java:
```diff
@@ -29,28 +31,43 @@
         // Create a scheduler to process scheduled tasks.
         final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
-
+    
```
Pointless whitespace change

xbreizh replied:
Removec

dermot-hardy commented on job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/ScheduledExecutor.java:
```diff
@@ -29,28 +31,43 @@
         scheduler.scheduleWithFixedDelay(task, 20, ScheduledExecutorConfig.getScheduledExecutorPeriod(),
                 TimeUnit.SECONDS);
-
+    
```
Pointless whitespace change

xbreizh replied:
Removed

dermot-hardy commented on job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/ScheduledExecutor.java:
```diff
@@ -29,28 +31,43 @@
+    public static void runAvailableJobs(final String origin)
+    {
+        try {
+            final Instant start = Instant.now();
+            DatabasePoller.pollDatabaseForJobsToRun();
+            final Instant end = Instant.now();
+            LOG.debug("Total time taken to execute scheduler for {} task in ms {}", origin, Duration.between(start, end).toMillis());
```
Consider using `isDebugEnabled()` when there is extra logic for the logging.

```suggestion
            if (LOG.isDebugEnabled()) {
                final Instant start = Instant.now();
                DatabasePoller.pollDatabaseForJobsToRun();
                final Instant end = Instant.now();
                LOG.debug("Total time taken to execute scheduler for {} task in ms {}", origin, Duration.between(start, end).toMillis());
            } else {
                DatabasePoller.pollDatabaseForJobsToRun();
            }
```

xbreizh replied:
Ok, I didn't know that one. Thx

dermot-hardy commented on job-service-acceptance-tests/src/test/java/com/hpe/caf/jobservice/acceptance/JobServiceEndToEndIT.java:
```diff
@@ -680,6 +680,7 @@ public void testCreateJobNoDelayNoPreReq() throws Exception {
         //retrieve job using web method
         final Job retrievedJob = jobsApi.getJob(partitionId, jobId, jobCorrelationId);
         LOG.info("--testCreateJobNoDelayNoPreReq job {} in partition: {}", retrievedJob.getId(), partitionId);
+        Thread.sleep(100); // Allows the scheduler to perform
```
You could consider removing this test altogether.  It's trying to check an internal implementation detail so it's not a good end-to-end test.

xbreizh replied:
Removed test

dermot-hardy commented on job-service-acceptance-tests/src/test/java/com/hpe/caf/jobservice/acceptance/JobServiceEndToEndIT.java:
```diff
@@ -706,6 +707,8 @@ public void testCreateJobNoDelayAndSomePreReq() throws Exception {
         LOG.info("--testCreateJobNoDelayAndSomePreReq job {} in partition: {}, canRun? {}", job1Id, partitionId, canRun);
         assertEquals(canRun, false, "Job "+job1Id+" is eligible to run despite incomplete prerequisite jobs");
 
+        Thread.sleep(100); // Allows the scheduler to perform
+
         JobServiceDatabaseUtil.assertJobTaskDataRowDoesNotExist(preReqJobId);
```
Same with this part of this test (i.e. the `final` assert that you've had to add a delay for).

xbreizh replied:
Removed test
